lib/url-decorator.jsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { CompositeDecorator } from 'draft-js';
+import urlRegex from 'url-regex';
+import normalizeUrl from 'normalize-url';
+import { viewExternalUrl } from './utils/url-utils';
+
+const urlRe = urlRegex();
+/*
+- markdownLinkRe is a regexp to be applied to the substring preceding a matched url, ie. `wholeString.slice(0, wholeString.indexOf(matchedUrl))`
+- it tests whether the matched url is preceded by the charcters that define a link in markdown, ie. `(` or `]: `
+*/
+const markdownLinkRe = /(?:\(|\]:\s+)$/;
+
+export function normalizeUrlsInMarkdownString( str ) {
+	return str.replace( urlRe, ( match, offset ) => {
+		var normalized = normalizeUrl( match );
+		if ( markdownLinkRe.test( str.slice( 0, offset ) ) ) {
+			return normalized;
+		}
+		// replace url with markdown link only if it's not already part of one
+		return `[${match}](${normalized})`;
+	} );
+}
+
+export const urlDecorator = new CompositeDecorator( [
+	{
+		strategy: function( contentBlock, callback ) {
+			const text = contentBlock.getText();
+			let matchArr, start;
+			while ( ( matchArr = urlRe.exec( text ) ) !== null ) {
+				start = matchArr.index;
+				callback( start, start + matchArr[0].length );
+			}
+		},
+
+		component: React.createClass( {
+			onClick: function( event ) {
+				event.preventDefault();
+				viewExternalUrl( normalizeUrl( this.props.decoratedText ) );
+			},
+
+			render: function() {
+				return (
+					<a
+						onClick={this.onClick}
+						href={this.props.decoratedText}>
+						{this.props.children}
+					</a>
+				);
+			}
+		} ),
```

Comment:
could you expand on why you'd like this component to be defined elsewhere?
do you mean to use it for something else as well?
if no, this does look to me like the clearest way of creating a decorator for a draft-js `Editor`, define the strategy and its component right next to each other (if not for any other reason because this way it's really clear what `decoratedText`is inside the component, it's exactly what you called the callback with in the strategy).

in terms of organisation I think what could be done is put `note-content-editor` and `url-decorator` in their own folder, `url-decorator` is only really meant to be used by `note-content-editor`. what do you think?